src/planner/partialize.c

```diff
@@ -185,6 +197,656 @@ partialize_agg_paths(RelOptInfo *rel)
 	return has_combine;
 }
 
+/*
+ * Get an an existing aggregation path for the given relation or NULL if no aggregation path exists.
+ */
+static bool
+has_min_max_agg_path(RelOptInfo *relation)
+{
+	ListCell *lc;
+	foreach (lc, relation->pathlist)
+	{
+		Path *path = lfirst(lc);
+		if (IsA(path, MinMaxAggPath))
+			return true;
+	}
+
+	return false;
+}
+
+/*
+ * Get an an existing aggregation path for the given relation or NULL if no aggregation path exists.
+ */
+static AggPath *
+get_existing_agg_path(RelOptInfo *relation)
+{
+	ListCell *lc;
+	foreach (lc, relation->pathlist)
+	{
+		Path *path = lfirst(lc);
+		if (IsA(path, AggPath))
+		{
+			AggPath *existing_agg_path = castNode(AggPath, path);
+			return existing_agg_path;
+		}
+	}
+
+	return NULL;
+}
+
+/*
+ * Get all subpaths from a Append, MergeAppend, or ChunkAppend path
+ */
+static List *
+get_subpaths_from_append_path(Path *path, bool handle_gather_path)
```

Comment:
When exactly do we want/not want to handle the Gather path? It's only for parallel paths, maybe we could just handle it always?

Reply:
This PR pushes down the aggregations for partial and non-partial paths. The `handle_gather_path` parameter is used to: (1) ensure we don't process gather nodes in non-partial paths, and (2) ensure the recursive call of this function does terminate.

Reply:
Just curious: why wouldn't it terminate without the boolean? Is there a loop of subpaths somewhere in the tree of nodes? Can a GatherPath have another GatherPath as child/subpath?

Reply:
Yes, it should terminate without the boolean, because we should never have a loop in the GatherPaths. However, I want to be on the safe side here and only descend the paths one GatherPaths at the most.